core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java

```diff
@@ -278,6 +278,37 @@ ClientRequestContext newDerivedContext(RequestId id, @Nullable HttpRequest req,
     @Nullable
     String fragment();
 
+    /**
+     * Returns the authority which will eventually be sent when a {@link Client} sends an {@link HttpRequest}.
+     * This method checks the following locations and returns the first non-null value.
+     * <ol>
+     *     <li>Either the {@link HttpHeaderNames#HOST} or {@link HttpHeaderNames#AUTHORITY} value from
+     *         {@link ClientRequestContext#additionalRequestHeaders()}.</li>
+     *     <li>The {@link HttpRequest#authority()} from {@link ClientRequestContext#request()}.</li>
+     *     <li>{@link ClientRequestContext#defaultRequestHeaders()}.</li>
+     *     <li>{@link Endpoint#authority()}.</li>
+     * </ol>
+     */
+    @Nullable
+    @UnstableApi
+    String authority();
+
+    /**
+     * Returns the uri constructed based on {@link ClientRequestContext#sessionProtocol()},
+     * {@link ClientRequestContext#authority()}, {@link ClientRequestContext#path()} and
+     * {@link ClientRequestContext#query()}.
+     *
+     * <p>Note that if the value of {@link ClientRequestContext#authority()}
+     * cannot be determined, the {@link String} "UNKNOWN" may be used as the authority.
```

Comment:
It is weird that `authority()` returns null for unknown but `uri()` returns a URI including `UNKNOWN`. Should we return a dummy URI? Is it unavoidable?

Reply:
This was one of the points I wasn't 100% sure about while working on this PR.

My original thought was to just match the previous behavior where `RequestHeaders#authority` was set to `UNKNOWN` if the endpoint was unknown.

https://github.com/line/armeria/blob/2c96fc350c6b123109d17009445d5071ceb4c414/core/src/main/java/com/linecorp/armeria/internal/client/DefaultClientRequestContext.java#L444

Having said this, I agree this can be confusing. I've just modified to return a `URI` with a `null` authority which I think makes more sense and causes less confusion.

Reply:
Note: an empty host/authority is deemed invalid is actually invalid for `http`. It seems like the jdk `URI` doesn't really check this restriction so I guess we're good for now.

> whereas the "http"
   scheme considers a missing authority or empty host invalid.

https://www.rfc-editor.org/rfc/rfc3986#section-3.2

Reply:
I didn't know you could create URIs with null permissions. It looks clearer than `UNKNOWN`.